lib/services/local/generic.go

```diff
@@ -0,0 +1,158 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// marshalFunc is a type signature for a marshaling function.
+type marshalFunc[T types.Resource] func(T, ...services.MarshalOption) ([]byte, error)
+
+// UnmarshalFunc is a type signature for an unmarshaling function.
+type unmarshalFunc[T types.Resource] func([]byte, ...services.MarshalOption) (T, error)
+
+// genericResourceService is a generic service for interacting with resources in the backend.
+type genericResourceService[T types.Resource] struct {
+	backend                   backend.Backend
+	resourceHumanReadableName string
+	limit                     int
+	backendPrefix             string
+	marshalFunc               marshalFunc[T]
+	unmarshalFunc             unmarshalFunc[T]
+}
+
+// ListResources returns a paginated list of resources.
+func (s *genericResourceService[T]) listResources(ctx context.Context, pageSize int, pageToken string) ([]T, string, error) {
+	rangeStart := backend.Key(s.backendPrefix, pageToken)
+	rangeEnd := backend.RangeEnd(rangeStart)
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > s.limit {
+		pageSize = s.limit
+	}
+
+	// If backend.NoLimit is requested and the service permits it, set the limit to NoLimit.
+	limit := pageSize + 1
+
+	// no filter provided get the range directly
+	result, err := s.backend.GetRange(ctx, rangeStart, rangeEnd, limit)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	out := make([]T, 0, len(result.Items))
+	for _, item := range result.Items {
+		resource, err := s.unmarshalFunc(item.Value)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		out = append(out, resource)
+	}
+
+	var nextKey string
+	if len(out) > pageSize {
+		nextKey = backend.GetPaginationKey(out[len(out)-1])
+		// Truncate the last item that was used to determine next row existence.
+		out = out[:pageSize]
+	}
+
+	return out, nextKey, nil
+}
+
+// getResource returns the specified resource.
+func (s *genericResourceService[T]) getResource(ctx context.Context, name string, extraKeyParts ...string) (resource T, err error) {
+	key := s.fullKey(name, extraKeyParts...)
+	item, err := s.backend.Get(ctx, key)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return resource, trace.NotFound("%s %q doesn't exist", s.resourceHumanReadableName, string(key))
```

Comment:
I've made a `displayName` function to make a display name for this that includes extra key parts if necessary. So it would be something along the lines of

```
SAMLIdPServiceProvider "provider1 (secrets/or/other/qualifier)" doesn't exist
```

Let me know if you're okay with that. I'm adding in these extraKeys thinking to situations like the above where there are other objects associated with a single object.

Reply:
Here's an example of one: https://github.com/gravitational/teleport/blob/master/lib/services/local/usertoken.go#L114

IMO I think this is a pretty low-effort way to support these use cases. It doesn't add much complexity.